data_structures/arrays/product_sum.py

```diff
@@ -0,0 +1,77 @@
+"""
+Calculate Product Sum from an Array.
+reference: https://dev.to/sfrasica/algorithms-product-sum-from-an-array-dc6
+
+For doctests run following command:
+python -m doctest -v product_sum.py
+
+We need to create a function that calculates the product sum of a "special" array.
+This array can contain integers or nested arrays. The product sum is obtained by
+adding all the elements together and multiplying by their respective depth.
+
+For example, in the array [x, y], the product sum is (x + y). In the array [x, [y, z]],
+the product sum is x + 2 * (y + z). In the array [x, [y, [z]]],
+the product sum is x + 2 * (y + 3z).
+
+Example Input:
+[5, 2, [-7, 1], 3, [6, [-13, 8], 4]]
+Output: 12
+
+"""
+
+
+def product_sum(arr: list[int | list], depth: int) -> int:
+    """
+    Recursively calculates the product sum of an array.
+
+    The product sum of an array is defined as the sum of its elements multiplied by
+    their respective depths.If an element is a list, its product sum is calculated
+    recursively by multiplying the sum of its elements with its depth plus one.
+
+    Args:
+        arr: The array of integers and nested lists.
+        depth: The current depth level.
+
+    Returns:
+        int: The product sum of the array.
+
+    Examples:
+        >>> product_sum([1, 2, 3], 1)
```

Comment:
Please add tests for:
```
>>> product_sum([1, 2, 3], -1)
>>> product_sum([1, 2, 3], 0)
>>> product_sum([1, 2, 3], 7)
```
Can you please provide one or more URLs to discussions of this algorithm so that visitors can read up on where and when it is useful?  This should also give us access to correct answers for our tests.